<h3><center>VWAP, Sharpe and Alpha Analysis<br>
CUNY SPS MSDS<br>
Professor Samuel Gralnick<br>
Team Sugar Cane: Euclid Zhang, Jie Zou, Zhenni Xie<br>
Fall 2022</center></h3>

In this analysis, we are performing calculations of the Volume-Weighted Average Price (VWAP), Sharpe Ratio, and Alpha of our investment portfolio and the individual assets in the portfolio, based on the asset prices from 8/25/2022 to 10/7/2022

In [422]:
#Loading packages

import numpy as np
from datetime import datetime, timedelta
import yfinance as yf
import pandas_datareader.data as pdr
import pandas as pd
import scipy.stats as stats
import math
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm

In [423]:
#The start date and end date of the assets' prices used for the calculations
start = datetime.strptime('2022-08-25', '%Y-%m-%d')
end = datetime.strptime('2022-10-08', '%Y-%m-%d')

Our investment portfolio includes the following assets, with purchasing prices (close prices) from 9/16/2022. In this analysis, **we will not include the U.S. 30 Year Treasury bonds in our calculation** because the metrics we are calculating are created to the compare portfolios of equities.

In [424]:
#Load the infromation of the assets in our portfolio
assets = pd.read_csv("portfolio assets.csv")
assets

Ticker                   Type                  Sector   Price      Units  \
0   US30Y  U.S. 30 Year Treasury                     NaN   90.44  467317.00   
1    HSON           Common Stock     COMMERCIAL SERVICES   34.95   36138.24   
2    GOGO           Common Stock          COMMUNICATIONS   13.32   94822.19   
3    JAKK           Common Stock       CONSUMER DURABLES   23.37   54044.99   
4    CALM           Common Stock   CONSUMER NON-DURABLES   56.85   22216.91   
5     HRB           Common Stock       CONSUMER SERVICES   44.93   28111.09   
6    HDSN           Common Stock   DISTRIBUTION SERVICES    7.94  159071.98   
7   BELFA           Common Stock   ELECTRONIC TECHNOLOGY   28.52   44285.81   
8    ARLP           Common Stock         ENERGY MINERALS   23.88   52890.76   
9      CI           Common Stock         HEALTH SERVICES  290.35    4350.03   
10   SRTS           Common Stock       HEALTH TECHNOLOGY   14.80   85339.97   
11    LNG           Common Stock     INDUSTRIAL SERVICES  167.25    7551.75   
12    BSM           Common Stock           MISCELLANEOUS   15.70   80447.87   
13   HUDI           Common Stock     NON-ENERGY MINERALS   29.64   42612.40   
14     CF           Common Stock      PROCESS INDUSTRIES   98.59   12810.95   
15    CSL           Common Stock  PRODUCER MANUFACTURING  291.30    4335.84   
16   MUSA           Common Stock            RETAIL TRADE  276.88    4561.65   
17   AZPN           Common Stock     TECHNOLOGY SERVICES  226.37    5579.50   
18    ASC           Common Stock          TRANSPORTATION   10.00  126303.15   
19     ED           Common Stock               UTILITIES   97.41   12966.13   
20    YCS                    ETF                     NaN   63.33   25261.96   
21    UUP                    ETF                     NaN   29.40   54416.32   
22    EUO                    ETF                     NaN   33.42   47870.73   
23    EWV                    ETF                     NaN   20.04   79832.33   
24    DIG                    ETF                     NaN   36.04   44390.67   
25    TTT                    ETF                     NaN   67.66   23645.28   
26    ERX                    ETF                     NaN   54.51   29349.47   
27    TMV                    ETF                     NaN  119.12   13430.49   
28    TBT                    ETF                     NaN   29.35   54509.02   
29    TYO                    ETF                     NaN   12.52  127782.74   

         Amount  
0   42262981.19  
1    1263031.49  
2    1263031.57  
3    1263031.42  
4    1263031.33  
5    1263031.27  
6    1263031.52  
7    1263031.30  
8    1263031.35  
9    1263031.21  
10   1263031.56  
11   1263030.19  
12   1263031.56  
13   1263031.54  
14   1263031.56  
15   1263030.19  
16   1263029.65  
17   1263031.42  
18   1263031.50  
19   1263030.72  
20   1599839.93  
21   1599839.81  
22   1599839.80  
23   1599839.89  
24   1599839.75  
25   1599839.64  
26   1599839.61  
27   1599839.97  
28   1599839.74  
29   1599839.90

In [425]:
#Calculate the weights of the assets in our portfolio
weights = np.array(assets['Amount'][1:]/sum(assets['Amount'][1:]))

We then retrieve the prices and volumes of the common stocks and ETFs in our portfolio, using the yfinance package

In [426]:
#retrieve the prices and volumes of the common stocks and ETFs in our portfolio, using the yfinance package

yf.pdr_override()

tickers = list(assets['Ticker'])[1:]

historical_data = pdr.DataReader(tickers, start, end , data_source='yahoo')
historical_data = historical_data.reset_index()
historical_data.Date = historical_data.Date.apply(datetime.date)
historical_data.set_index('Date', inplace = True)


[*********************100%***********************]  29 of 29 completed


First we calculate the Volume-Weighted Average Price (VWAP) using the following formula

<TABLE STYLE="display:inline-table; font-size:100%; border-collapse:collapse; vertical-align:middle"><TR><TD STYLE="padding:0px; border-bottom: solid 1px; width:500px; height: 20px; text-align:center"><span>&Sigma;</span> [(daily high price + daily low price + daily close price) * daily trading volume]</TD></TR> <TR><TD STYLE="padding:0px; text-align:center"><span>&Sigma;</span> daily trading volume</TD></TR></TABLE>

In [427]:
total_traded = ((historical_data['High'] + historical_data['Low'] + historical_data['Close'])/3)*historical_data['Volume']
total_traded.fillna(0, inplace = True)
VWAP = pd.DataFrame(total_traded.sum() / historical_data['Volume'].sum())
VWAP.columns = ['VWAP']

In [428]:
VWAP = VWAP.merge(right = assets[['Ticker','Price']], how = "inner", left_index = True, right_on="Ticker").set_index('Ticker')
VWAP = VWAP.merge(pd.DataFrame(historical_data['Adj Close'].iloc[historical_data['Adj Close'].shape[0]-1,]),
                 how = "inner", left_index = True, right_index = True)
VWAP.rename(columns={'Price': 'Purchased Price'}, inplace=True)

#adding VWAP for the portfolio
port_VWAP = pd.DataFrame(VWAP.T.dot(weights)).T
port_VWAP.index = ["Portfolio"]
VWAP = round(VWAP.append(port_VWAP),2)
VWAP.tail(5)

VWAP  Purchased Price  2022-10-07
TTT        69.32            67.66       78.91
TYO        12.78            12.52       13.84
UUP        29.78            29.40       30.30
YCS        63.35            63.33       65.74
Portfolio  76.93            76.79       78.40

Next, we calculate the Sharpe Ratio using the following formula

<TABLE STYLE="display:inline-table; font-size:100%; border-collapse:collapse; vertical-align:middle"><TR><TD STYLE="padding:0px; border-bottom: solid 1px; width:500px; height: 20px; text-align:center">return of portfolio - risk-free rate</TD></TR> <TR><TD STYLE="padding:0px; text-align:center">standard deviation of the portfolio's return</TD></TR></TABLE>

For the risk-free rate, we use the yield rate of the U.S. 3 Month Treasury Bill as of 10/07/2022, 3.365% (https://www.marketwatch.com/investing/bond/tmubmusd03m?countrycode=bx). We calculated the daily return using 251 trade days in a year. The daily gain is

In [429]:
risk_free_rate = 1.03365**(1/251) -1
risk_free_rate

0.0001318661736162685

We then calculate the Sharpe Ratios using the daily Adjusted Close Price

In [430]:
close_prices = historical_data['Adj Close']
returns = close_prices.pct_change().dropna()
returns['Portfolio'] = returns.dot(weights)
sharp_ratios = pd.DataFrame((returns.mean() - risk_free_rate) / returns.std())
sharp_ratios.columns = ['Sharpe Ratio']

In [431]:
sharp_ratios.tail(5)

Sharpe Ratio
TTT            0.258891
TYO            0.291776
UUP            0.186173
YCS            0.317407
Portfolio      0.040293

Now we calcuate the alpha of the portfolio and assets, which is derived from the following model

return of investment - risk-free rate = alpha + beta * (return of benchmark - risk-free rate) + random error

In this analysis, we calculate three different alphas using the S&P 500 ("^GSPC"), Dow Jones ("^DJI"), and Nasdaq Composite ("^IXIC") indexes as benchmarks. Instead of using the current risk-free rate, we use the historical yield rates (U.S. 3 Month Treasury Bill) within the same time frame to calculate the alphas.

In [432]:
#load the U.S. 3 Month Treasury Bill yield rates, downloaded from 
#https://www.marketwatch.com/investing/bond/tmubmusd03m?countrycode=bx

risk_free_rates = pd.read_csv("BOND_BX_XTUP_TMUBMUSD03M.csv")
risk_free_rates.Date  = pd.to_datetime(risk_free_rates.Date)
risk_free_rates.set_index('Date', inplace = True)
risk_free_rates.sort_index(inplace = True)
risk_free_rates['Close']  = risk_free_rates['Close'].str.rstrip("%").astype("float") / 100
risk_free_rates['Close'] = (1 + risk_free_rates['Close'])**(1/251)-1
risk_free_rates = risk_free_rates[str(start.date()):str(end.date())]['Close']

In [433]:
#retrieve the prices of the market indexes, using the yfinance package

tickers = tickers = ["^GSPC", "^DJI", "^IXIC"]
index_prices = pdr.DataReader(tickers, start, end , data_source='yahoo')
index_prices = index_prices['Adj Close']

index_prices = index_prices.reset_index()
index_prices.Date = index_prices.Date.apply(datetime.date)
index_prices.set_index('Date', inplace = True)

#calcuate the daily percent returns of the indexes
market_rates = index_prices.pct_change().dropna()

[*********************100%***********************]  3 of 3 completed


In [434]:
#define a function to calculate the alphas. Addtionally, we can get the beta and R-squared from the results.

#The code is from the following post
#https://firemymoneymanager.com/getting-started-using-python-to-find-alpha/
def market_compare(market_ticker):
    #new dataframe to store the results of the calculations
    regressions = pd.DataFrame(columns = ['ticker', 'alpha', 'beta', 'rsquared'])

    for i in range(0, len(returns.columns)):
        # for calculating y, we must subtract the risk free rate
        y = (returns[returns.columns[i]] - risk_free_rates).dropna()
        X = (market_rates[market_ticker] - risk_free_rates).dropna()
        X = sm.add_constant(X)
        X.columns = ['const','Mkt-RF']

        # training the model
        model = sm.OLS(y, X).fit()

        # append the results
        regressions.loc[i] = [ 
            returns.columns[i],
            model.params['const'],
            model.params['Mkt-RF'],
            model.rsquared
        ]
    regressions.set_index('ticker', inplace = True)
    return regressions

Calculate the alphas using three different market indexes

In [435]:
S_P500_Compare = market_compare('^GSPC')
Dow_Jones_Compare = market_compare('^DJI')
Nasdaq_Compare = market_compare('^IXIC')

The following shows some of the results from the calculation using S&P 500 index as the benchmark

In [436]:
S_P500_Compare.tail(5)

alpha      beta  rsquared
ticker                                 
TTT        0.008250 -0.504184  0.042376
TYO        0.003927 -0.572935  0.178468
UUP       -0.000043 -0.291778  0.467197
YCS        0.002917 -0.284953  0.126693
Portfolio  0.002527  0.429578  0.362868

Fainlly, combine the reuslt of all calculations together so we can compare the numbers

In [437]:
summary = pd.DataFrame(S_P500_Compare['alpha']).merge(pd.DataFrame(Dow_Jones_Compare['alpha']), how = "inner", left_index = True, right_index = True)
summary = summary.merge(pd.DataFrame(Nasdaq_Compare['alpha']), how = "inner", left_index = True, right_index = True)
summary.columns = ['alpha_SP500', 'alpha_Dow_Jones', 'alpha_Nasdaq']
summary = sharp_ratios.merge(summary, how = "inner", left_index = True, right_index = True)
summary = VWAP.merge(summary, how = "inner", left_index = True, right_index = True)
summary

VWAP  Purchased Price  2022-10-07  Sharpe Ratio  alpha_SP500  \
ARLP        24.59            23.88       24.31     -0.037640     0.002477   
ASC          9.72            10.00        9.53     -0.005351     0.000966   
AZPN       228.50           226.37      253.41      0.278757     0.008350   
BELFA       29.62            28.52       29.03     -0.105511    -0.001059   
BSM         15.97            15.70       17.36      0.106568     0.005186   
CALM        57.21            56.85       53.63     -0.031122     0.001702   
CF         102.63            98.59      104.80     -0.093633     0.000446   
CI         286.76           290.35      288.38     -0.022557     0.002893   
CSL        294.12           291.30      292.45     -0.137457     0.001818   
DIG         35.23            36.04       38.81     -0.009746     0.008873   
ED          94.15            97.41       81.58     -0.368656    -0.002916   
ERX         55.00            54.51       59.27     -0.009438     0.008908   
EUO         34.34            33.42       35.28      0.104874    -0.001111   
EWV         20.58            20.04       21.42      0.257220     0.000566   
GOGO        13.11            13.32       12.71     -0.269087    -0.004542   
HDSN         8.17             7.94        7.52     -0.254706    -0.004986   
HRB         43.78            44.93       38.05     -0.323418    -0.004123   
HSON        34.68            34.95       32.57     -0.241838    -0.003422   
HUDI        23.48            29.64       25.09      0.198391     0.022074   
JAKK        22.78            23.37       19.88     -0.193696    -0.001289   
LNG        164.29           167.25      174.00      0.037121     0.002856   
MUSA       281.62           276.88      269.14     -0.185857    -0.000360   
SRTS        13.05            14.80       11.93     -0.086363     0.000191   
TBT         29.77            29.35       32.60      0.257472     0.005416   
TMV        122.16           119.12      138.59      0.258431     0.008075   
TTT         69.32            67.66       78.91      0.258891     0.008250   
TYO         12.78            12.52       13.84      0.291776     0.003927   
UUP         29.78            29.40       30.30      0.186173    -0.000043   
YCS         63.35            63.33       65.74      0.317407     0.002917   
Portfolio   76.93            76.79       78.40      0.040293     0.002527   

           alpha_Dow_Jones  alpha_Nasdaq  
ARLP              0.002153      0.002812  
ASC               0.000789      0.001041  
AZPN              0.008068      0.008586  
BELFA            -0.001097     -0.000781  
BSM               0.005072      0.005357  
CALM              0.001442      0.001986  
CF                0.000400      0.000516  
CI                0.003031      0.002774  
CSL               0.001723      0.001823  
DIG               0.008822      0.008746  
ED               -0.002964     -0.003033  
ERX               0.008876      0.008772  
EUO              -0.001157     -0.001083  
EWV               0.000408      0.000542  
GOGO             -0.004563     -0.004492  
HDSN             -0.005030     -0.004891  
HRB              -0.004386     -0.003896  
HSON             -0.003121     -0.003893  
HUDI              0.021163      0.022817  
JAKK             -0.001708     -0.000933  
LNG               0.002854      0.002778  
MUSA             -0.000268     -0.000545  
SRTS             -0.000218      0.000397  
TBT               0.005410      0.005470  
TMV               0.008051      0.008168  
TTT               0.008234      0.008341  
TYO               0.003936      0.004014  
UUP              -0.000070     -0.000018  
YCS               0.002848      0.002791  
Portfolio         0.002414      0.002599

## Conclusion

We purchased the portfolio with a price (76.79 per unit) lower than the VWAP (76.93 per unit) and the current value of a unit is 78.40. We are making a profit and we may consider selling some assets that the current value is higher than the VWAP. The Sharpe Ratio is positive, indicating that our equities has a better return than the risk-free investment. Some of the individual assets have a negative Sharpe Ratio, we may sell some of them and reallocate the fund to risk-free investments or other equities. The Alphas calculated using three different benchmarks are showing that our portfolio outperformed the three market indexes. Overall, our portfolio is performing well. Some minor adjustments may be good, but we can hold the assets for now as the observation is based on a few weeks only.